File: src/mobiletransformers/artifacts/package_paths.py
"""One resolver for every stage path in a package. No consumer appends a stage name to a string.

## Why this exists

A package has **two** on-disk layouts, and until now both were built by string concatenation at every
call site:

===================  ==========================================  ==========================
layout               shape                                        who produces it
===================  ==========================================  ==========================
hub package          ``variants/<variantId>/{train,inference,     ``export/pipeline.py``
                     embedding}`` + ``shared/tokenizer``
device cache (flat)  ``<cacheDir>/<repoId>/{train,inference,      ``ModelPackageInstaller``,
                     embedding,tokenizer}``                        ``scripts/device_package.sh``
===================  ==========================================  ==========================

The manifest has always *declared* the hub layout — ``variant.paths`` — but exactly one consumer in the
whole repo read it (``cli/federated.py``); roughly twenty others spelled the join by hand, in four
languages. The #35 simulation lost a cycle to precisely this: the client looked for ``<package>/train/``
(the cache layout) in a hub package, and ORT reported ``INVALID_ARGUMENT : Invalid fd was supplied: -1``,
naming no file.

This is the layer-identity problem in a second namespace. The fix is the same one ``cpp/layer_name.h``
applied there: **one place that knows the spelling**, and a guard that keeps it that way.

## The rule

Ask a :class:`PackagePaths` for a stage. Never write ``dir / "train"``.

Mirrored in Kotlin by ``packages/PackagePaths.kt``; the two must agree, because the same package is
read by both.

**C++ deliberately has no mirror.** It never resolves a stage — Kotlin hands it an already-resolved
directory over JNI, and its joins (``inference_dir + "/weight_handoff_map.json"``) append a *filename*
to a resolved directory, which is not this defect. Adding a third copy for symmetry would create the
very duplication this module exists to remove.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from mobiletransformers.exceptions import ManifestError

#: Stage keys as they appear in the manifest's ``paths`` map. These are wire names.
STAGE_INFERENCE = "inference"
STAGE_TRAIN = "train"
STAGE_EMBEDDING = "embedding"
STAGE_TOKENIZER = "tokenizer"

#: Every stage a package can declare, in a deterministic order.
STAGES: tuple[str, ...] = (STAGE_INFERENCE, STAGE_TRAIN, STAGE_EMBEDDING, STAGE_TOKENIZER)

#: Directory name each stage takes in the FLAT device-cache layout.
#:
#: Note ``embedding`` keeps its name while ``tokenizer`` moves from ``shared/tokenizer`` to a sibling —
#: the cache layout is not simply the hub layout with the ``variants/<id>/`` prefix removed, which is
#: exactly why open-coding it in nine places produced two different answers.
_CACHE_DIRNAMES: dict[str, str] = {
    STAGE_INFERENCE: "inference",
    STAGE_TRAIN: "train",
    STAGE_EMBEDDING: "embedding",
    STAGE_TOKENIZER: "tokenizer",
}

#: Filename of the weight handoff map inside the inference stage.
WEIGHT_HANDOFF_FILENAME = "weight_handoff_map.json"


@dataclass(frozen=True)
class PackagePaths:
    """Resolved absolute stage directories for one package in one layout.

    Build with :meth:`for_hub` or :meth:`for_cache`; do not construct directly unless you are writing a
    third layout, in which case add a factory here rather than joining strings at the call site.
    """

    root: Path
    #: Stage key -> absolute directory. Only stages the layout actually declares are present.
    stages: dict[str, Path]
    #: Which layout produced this, for error messages that need to say so.
    layout: str

    @classmethod
    def for_hub(cls, package_dir: str | Path, variant: object) -> PackagePaths:
        """Resolve against a hub package using the variant's **declared** ``paths``.

        :param variant: a ``SelectedVariant`` (anything exposing a ``paths`` mapping). The manifest is
            the source of truth here — a variant may legitimately place a stage somewhere this module
            would not have guessed, and re-deriving ``variants/<id>/<stage>`` would silently ignore it.
        """
        package_dir = Path(package_dir)
        declared = getattr(variant, "paths", None)
        if not isinstance(declared, dict):
            raise ManifestError(
                "variant declares no `paths` map; a package built before the manifest carried per-variant "
                "paths cannot be resolved — re-export it."
            )
        stages = {stage: package_dir / rel for stage, rel in declared.items() if isinstance(rel, str) and rel}
        return cls(root=package_dir, stages=stages, layout="hub")

    @classmethod
    def for_cache(cls, cache_dir: str | Path, repo_id: str) -> PackagePaths:
        """Resolve against the FLAT on-device cache layout.

        ``repo_id`` must already be sanitized (``/`` -> ``__``) if it came from a hub id; that mapping is
        owned by ``hub/package_format.py::sanitize_repo_id`` and mirrored in Kotlin, and is deliberately
        not repeated here.
        """
        base = Path(cache_dir) / repo_id
        return cls(
            root=base,
            stages={stage: base / name for stage, name in _CACHE_DIRNAMES.items()},
            layout="cache",
        )

    # -- accessors ----------------------------------------------------------

    def stage(self, name: str) -> Path:
        """The directory for ``name``, or :class:`ManifestError` naming what is available.

        Fails closed rather than returning a plausible path that does not exist: a silently-wrong stage
        directory surfaces later as an unrelated-looking IO error, which is the failure mode this module
        was written to end.
        """
        if name not in STAGES:
            raise ManifestError(f"unknown stage {name!r}; known stages are {list(STAGES)}")
        try:
            return self.stages[name]
        except KeyError:
            raise ManifestError(
                f"this {self.layout} package does not declare a {name!r} stage "
                f"(declared: {sorted(self.stages)})"
            ) from None

    @property
    def inference(self) -> Path:
        return self.stage(STAGE_INFERENCE)

    @property
    def train(self) -> Path:
        return self.stage(STAGE_TRAIN)

    @property
    def embedding(self) -> Path:
        return self.stage(STAGE_EMBEDDING)

    @property
    def tokenizer(self) -> Path:
        return self.stage(STAGE_TOKENIZER)

    @property
    def weight_handoff(self) -> Path:
        """The handoff map, which lives inside the inference stage in both layouts."""
        return self.inference / WEIGHT_HANDOFF_FILENAME

    def has(self, name: str) -> bool:
        """Whether the layout declares ``name`` at all (says nothing about what is on disk)."""
        return name in self.stages

File: android/MobileTransformers/MobileTransformers/src/main/java/com/martinkorelic/mobiletransformers/packages/PackagePaths.kt
package com.martinkorelic.mobiletransformers.packages

import java.io.File

/**
 * One resolver for every stage directory. **No consumer builds a stage path by appending a string.**
 *
 * Kotlin mirror of `artifacts/package_paths.py`; `cpp/package_paths.h` is the third. All three read the
 * same package, so all three must agree.
 *
 * Two layouts exist and they are NOT the same shape:
 *
 * | layout        | shape                                                              |
 * |---------------|--------------------------------------------------------------------|
 * | hub package   | `variants/<variantId>/{train,inference,embedding}` + `shared/tokenizer` |
 * | device cache  | `<cacheDir>/<repoId>/{train,inference,embedding,tokenizer}` (FLAT)  |
 *
 * The manifest has always declared the hub layout in `variant.paths`, and on the Kotlin side **nothing
 * read those values** — `ManifestValidator` only checked that the keys were present, while nine call
 * sites spelled `"$cacheDir/$repoId/inference"` by hand. `TrainingWorker` carried a comment marking
 * itself as "one place that knows the cache layout"; this is now that place, for everyone.
 *
 * This is the layer-identity problem in a second namespace, and it gets the same structural answer
 * `layer_name.h` gave the first: one spelling, one owner, a guard to keep it.
 */
class PackagePaths private constructor(
    val root: File,
    private val stages: Map<String, File>,
    private val layout: String,
) {

    companion object {
        const val STAGE_INFERENCE = "inference"
        const val STAGE_TRAIN = "train"
        const val STAGE_EMBEDDING = "embedding"
        const val STAGE_TOKENIZER = "tokenizer"

        val STAGES = listOf(STAGE_INFERENCE, STAGE_TRAIN, STAGE_EMBEDDING, STAGE_TOKENIZER)

        const val WEIGHT_HANDOFF_FILENAME = "weight_handoff_map.json"

        /** The ObjectBox store's directory name, inside the embedding stage. */
        const val EMBEDDING_DATABASE_DIRNAME = "database"

        /**
         * Resolve against a hub package using the variant's DECLARED paths.
         *
         * The manifest decides. A variant may legitimately place a stage somewhere this class would not
         * guess — `tokenizer` already lives at `shared/tokenizer` rather than under `variants/<id>/` —
         * so re-deriving the convention would quietly ignore the declaration.
         */
        @JvmStatic
        fun forHub(packageDir: File, variant: MobileTransformersManifest.Variant): PackagePaths {
            val declared = variant.paths
            require(declared.isNotEmpty()) {
                "variant '${variant.id}' declares no `paths`; a package built before the manifest " +
                    "carried per-variant paths cannot be resolved — re-export it."
            }
            val stages = declared
                .filterValues { it.isNotBlank() }
                .mapValues { (_, rel) -> File(packageDir, rel) }
            return PackagePaths(packageDir, stages, layout = "hub")
        }

        /**
         * Resolve against the FLAT on-device cache layout.
         *
         * [repoId] must already be sanitized — that mapping belongs to [PackageFormat.sanitizeRepoId]
         * and is deliberately not repeated here.
         */
        @JvmStatic
        fun forCache(cacheDir: File, repoId: String): PackagePaths {
            val base = File(cacheDir, repoId)
            return PackagePaths(
                root = base,
                stages = STAGES.associateWith { File(base, it) },
                layout = "cache",
            )
        }

        /** Convenience for the many call sites that hold the cache dir as a string. */
        @JvmStatic
        fun forCache(cacheDir: String, repoId: String): PackagePaths = forCache(File(cacheDir), repoId)
    }

    /**
     * The directory for [name], or [IllegalArgumentException] naming what this layout does declare.
     *
     * Fails closed rather than handing back a plausible path that does not exist: a silently-wrong
     * stage surfaces much later as an unrelated-looking IO error — the #35 client's
     * `INVALID_ARGUMENT : Invalid fd was supplied: -1`, which named no file at all.
     */
    fun stage(name: String): File {
        require(name in STAGES) { "unknown stage '$name'; known stages are $STAGES" }
        return stages[name] ?: throw IllegalArgumentException(
            "this $layout package does not declare a '$name' stage (declared: ${stages.keys.sorted()})"
        )
    }

    val inference: File get() = stage(STAGE_INFERENCE)
    val train: File get() = stage(STAGE_TRAIN)
    val embedding: File get() = stage(STAGE_EMBEDDING)
    val tokenizer: File get() = stage(STAGE_TOKENIZER)

    /** The handoff map, which lives inside the inference stage in both layouts. */
    val weightHandoff: File get() = File(inference, WEIGHT_HANDOFF_FILENAME)

    /**
     * The RAG vector store, INSIDE the embedding stage.
     *
     * Not shipped — ingestion creates it — which is why the two RAG sites that used to spell
     * `"$cacheDir/$repoId/embedding/database"` were carried as guard debt rather than exempted. A
     * sub-path of a stage still has to start from a resolved stage, or it drifts the same way a stage
     * does: the retriever's own comment claimed `cacheDir/modelName/database/` while the code wrote
     * `embedding/database/`.
     */
    val embeddingDatabase: File get() = File(embedding, EMBEDDING_DATABASE_DIRNAME)

    /**
     * The EMBEDDER's tokenizer, inside the embedding stage — a different tokenizer from
     * [tokenizer], which belongs to the generation model.
     */
    val embeddingTokenizer: File get() = File(embedding, STAGE_TOKENIZER)

    /** Whether the layout declares [name] at all (says nothing about what is on disk). */
    fun has(name: String): Boolean = stages.containsKey(name)
}
